### ISSUE TITLE
Feature Request: Sync Categories

### ISSUE THREAD

**Issue by tensau** [dmfs/opentasks#14]

**Comment by avanc** [comment]
I wonder that there is not more interest in this feature. How are other people organize all different kind of tasks? E.g.: I don't want to see all tasks related to computer, if I'm on a train without my laptop.

owncloud will get support for categories soon: https://github.com/owncloud/tasks/pull/172
With categories support in this app, I can get rid of my current propietary todo list.

So, I strongly vote for this feature :+1:

**Comment by dmfs** [comment]
Our task back-end already supports categories. The sync part is almost done too. We just have to work on the UI.

**Comment by avanc** [comment]
Is this feature still on the roadmap?

**Comment by dmfs** [comment]
Yes. We're working on it. We've been quite busy with paid projects, but I hope we'll have more time for this soon.

**Comment by avanc** [comment]
Just a ping so this problem is not forgotten :-)

Currently, I share my calendar including the tasks with my girlfriend. We have one task as our shopping list and add the items into this task, as we don't want to have our task list cluttered with a lot of grocery items. However, this gives often problems on sync as we both added items to the shopping list. As conflicts within one task are not handled, some items are lost.

**Comment by dmfs** [comment]
Sorry, this is not forgotten, we just had to revisit priorities, for business reasons.

**Comment by censlin** [comment]
Would it be possible to give the categories colours? E.g. letting the category change the background colour of a task in the widget. I would imagine that it won't be possible to sync the colours with a server, but I'm ok with manually setting the colours on each client.

**Comment by jkufner** [comment]
@censlin Colors are typically used to represent calendar where the task/event belongs.

**Comment by censlin** [comment]
Yeah, I figured that the calendar's colour would still be displayed in the box to the left of the task as it is now. The category task would be displayed by a separate mechanisms, like the background colour or a coloured border around the box.

**Comment by censlin** [comment]
Currently I use separate calendars for each project, so definitely still want calendar colours

**Comment by jkufner** [comment]
Then it would make sense to display categories as 'tags'. Just like many other apps do, using colored label, like Gitlab and Github do in issue lists.

**Comment by censlin** [comment]
Yeah, that would do the job for me.

**Comment by censlin** [comment]
It might cause some space issues for some though, if the tags have text. In my case it just means stretching the widget from 3 columns to 4 columns. But I'm not sure if that can be done on smaller screens.

**Comment by mcronce** [comment]
@dmfs Where does this stand currently?  I saw further up that the back end work was already partially done and partially in progress and that this item would really just be the required UI changes.  Is that still the case?  If so, are you open to pull requests? ;)

**Comment by macmeck** [comment]
+1 for categories!

**Comment by nicorikken** [comment]
@dmfs you mentioned some work was already done, is that availabe on a branch somewhere? Perhaps I can join in.

I need some more organisation in my tasks, and categories seems my best solution, instead of separate tasklists for my projects, or switching to Taskdroid. Even just a comma separated read-only listing would be a worthwile improvement.

**Comment by avanc** [comment]
@nicorikken Is Taskdroid capable to use the the tasks synced with caldav sync?

I'm still confused how Android handles tasks or events. Caldav syncs syncs events to the Android device. Thus, every calendar app can use this data. Is the same not valid for tasks? So I would expect a lot of task applications around. And even a native one similar to the native calendar app...

**Comment by korelstar** [comment]
@avanc 
Android provides an API ([Content Provider](https://developer.android.com/guide/topics/providers/content-providers.html)) for calendars containing events. Hence, every calendar app can use this data. However, there is no official API for tasks. Therefore, opentasks (i.e. [opentasks-provider](https://github.com/dmfs/opentasks-provider)) has introduced an own Content Provider (API) for this. Since it's *open*, every app can use the data, but the maintainer has to integrate it manually, since it's not part of the Android SDK. Therefore, most of the calendar apps don't support tasks and there is no *native* app for tasks. However, you can ask Google to directly integrate opentasks into Android :grin:

**Comment by tobixen** [comment]
+1

I'm using the categories field actively, caldav and syncing the tasks to my mobile.

Worst of all, when editing todo tasks from my mobile, the categories field in the VTODO is blanked out.  (I suspect this is a general bug, that all fields that aren't supported gets blanked out, probably I should do some research on this and report it as a separate issue)

**Comment by tobixen** [comment]
> We've been quite busy with paid projects

Eventually, how much would it cost to get this one prioritized? :-)

**Comment by kedront** [comment]
Any new plans to work on the UI? I need category sync too.
Can we spent some money to prioritize this task?

**Comment by L3R0Y** [comment]
+1 for categories here as well!
I am open for a donation or might even help to code it.

**Comment by Noctiflore** [comment]
+1. At least OpenTasks should not delete categories set by other clients, in case of a CalDAV share for example.

**Comment by vauvenal5** [comment]
Any news here?

**Comment by bjuergens** [comment]
I'd like to help with this. 

* Is there existing code for this / Is there a branch, where some work had been done already? 
* As I understand it, the back-end already works, and we "only" need to a support for the front-end. Is this correct?

**Comment by dmfs** [comment]
Yes, I believe the back-end should be ready to do. We certainly could use some integration tests to confirm this. This should be much easier now that we have the opentaskspal module.

We don't have anything for the UI yet. However, the [subtask branch](https://github.com/dmfs/opentasks/tree/stories/442-show-subtasks) should get you going. Just like subtasks, categories should be implemented in the `ViewFragment` first and use RX for async operations.

**Comment by christf** [comment]
The most pressing problem seems to be that categories are currently destroyed when editing. Is there a workaround for that?

**Comment by psxde** [comment]
+1 at least for not removing existing categories when editing a task

**Comment by avanc** [comment]
Any news on this. If I got it right, it is _just_ an HMI issue as the sync adapter already supports categories.

**Comment by dmfs** [comment]
The task provider has support for categories. However, as mentioned before the code is pretty much untested. Especially with the recurrence handling we have now, we should create tests to ensure categories are properly maintained when creating task overrides.
Other than that, it's only the UI that's missing for categories support in this UI.

**Comment by jhamb** [comment]
+1 for showing Task Categories in Android app

**Comment by wToad** [comment]
Hi, it seems the UI for categories is missing for few years now. But it seems to be a very popular feature, why isn't it more prioritized than that?

**Comment by dmfs** [comment]
Mostly because recurrence was an even more popular feature IMO and that's where most of the work was spent, apart from bug fixes.

**Comment by tobixen** [comment]
Recurrent tasks?  That's interesting.

I was working a bit with it in my calendar-cli client tool - many tasks are repetitive by nature - for instance, one would maybe want to "clean the floor" weekly.  The RFC apparently supports a recurrent task - but as far as I could find, the RFC didn't describe very well how to mark this weeks cleaning as "completed".  In calendar-cli I solved it by forking the vtodo item on the client side into an instance-object which got marked "completed", and another vtodo for the series which was not completed.  But I digress ...

**Comment by dmfs** [comment]
@tobixen that's essentially what we do too. The "RFC-way" would be to create an overridden instance in the same VCalendar object. But every completed instance would inflate the original object for no good reason. Splitting the completed instance off seems a reasonable solution.

**Comment by Zvezdin** [comment]
Any update on this? Otherwise, does anyone know an Android CalDav tasks frontend which supports categories and possibly subtasks?

**Comment by thomas-mc-work** [comment]
Yes, this one: https://f-droid.org/en/packages/org.tasks/

**Comment by rgloor** [comment]
+1 for categories (and appropriate filters, sorting)

**Comment by rgloor** [comment]
....is there presently any work done on this issue?
Thanks in advance for any feedback.

**Comment by bjuergens** [comment]
I am not working on this. I had a look at it and soon noticed, that I don't know as much about android-development than I thought I did
